br/pkg/restore/client.go

```diff
@@ -1050,39 +1087,58 @@ func (rc *Client) switchTiKVMode(ctx context.Context, mode import_sstpb.SwitchMo
 	}
 	bfConf := backoff.DefaultConfig
 	bfConf.MaxDelay = time.Second * 3
+
+	eg, ectx := errgroup.WithContext(ctx)
+
+CommunicateWithAllStores:
 	for _, store := range stores {
-		opt := grpc.WithInsecure()
-		if rc.tlsConf != nil {
-			opt = grpc.WithTransportCredentials(credentials.NewTLS(rc.tlsConf))
-		}
-		gctx, cancel := context.WithTimeout(ctx, time.Second*5)
-		connection, err := grpc.DialContext(
-			gctx,
-			store.GetAddress(),
-			opt,
-			grpc.WithBlock(),
-			grpc.FailOnNonTempDialError(true),
-			grpc.WithConnectParams(grpc.ConnectParams{Backoff: bfConf}),
-			// we don't need to set keepalive timeout here, because the connection lives
-			// at most 5s. (shorter than minimal value for keepalive time!)
-		)
-		cancel()
-		if err != nil {
-			return errors.Trace(err)
-		}
-		client := import_sstpb.NewImportSSTClient(connection)
-		_, err = client.SwitchMode(ctx, &import_sstpb.SwitchModeRequest{
-			Mode: mode,
-		})
-		if err != nil {
-			return errors.Trace(err)
-		}
-		err = connection.Close()
-		if err != nil {
-			log.Error("close grpc connection failed in switch mode", zap.Error(err))
-			continue
+		select {
+		case <-ectx.Done():
+			// when got the first error from eg, ectx will be canceled and then break the loop
+			break CommunicateWithAllStores
```

Comment:
I guess just test `ectx.Err() != nil` would be enough?

Reply:
You mean remove the `select{...}` and then test `ectx.Err() != nil` in each round of the `for _, store := range stores` loop?

Reply:
> You mean remove the `select{...}` and then test `ectx.Err() != nil` in each round of the `for _, store := range stores` loop?

Yes

Reply:
According to the document, the return value of `Err()` implies the state of the context:
```go
	// If Done is not yet closed, Err returns nil.
	// If Done is closed, Err returns a non-nil error explaining why:
	// Canceled if the context was canceled
	// or DeadlineExceeded if the context's deadline passed.
	// After Err returns a non-nil error, successive calls to Err return the same error.
	[Err](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/context/context.go;bpv=1;bpt=1;l=106?gsn=Err&gs=kythe%3A%2F%2Fgo.googlesource.com%2Fgo%3Flang%3Dgo%3Fpath%3Dcontext%23method%2520Context.Err)() error
```

Reply:
You are right, testing `Err()` instead of using `select` might be a simpler implementation. Now I have modified it. PTLA